src/pybind/mgr/cephadm/serve.py

```diff
@@ -742,8 +828,19 @@ def _apply_service(self, spec: ServiceSpec) -> bool:
             # TODO: return True would result in a busy loop
             # can't know if daemon count changed; create_from_spec doesn't
             # return a solid indication
-            return False
+            return None
 
+        try:
+            slots_to_add, daemons_to_remove, rank_map = self.discover_daemons_to_add_and_remove_by_service(spec)
+            self.mgr.daemon_deploy_queue.add_to_queue([(daemon_to_add, spec) for daemon_to_add in slots_to_add])
+            self.mgr.daemon_removal_queue.add_to_queue([(daemon_to_remove, spec) for daemon_to_remove in daemons_to_remove])
```

Comment:
nit: I think slots_to_xxx have already the operation so there's no need to repeat the operation in the iterator variable as well for better readability. So we can have something like:

```suggestion
            self.mgr.daemon_deploy_queue.add_to_queue([(daemon, spec) for daemon in slots_to_add])
            self.mgr.daemon_removal_queue.add_to_queue([(daemon, spec) for daemon in daemons_to_remove])
```